README.md
# inSilecoRef
[![R CMD Check](https://github.com/inSileco/inSilecoRef/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/inSileco/inSilecoRef/actions/workflows/R-CMD-check.yaml)
[![pkgdown](https://github.com/inSileco/inSilecoRef/actions/workflows/pkgdown.yaml/badge.svg)](https://github.com/inSileco/inSilecoRef/actions/workflows/pkgdown.yaml)
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

# Description

This package contains the tools used to manage bibliographic data for the
[*inSileco* blog](https://insileco.github.io/).

Managing bibliographies in a group can be hard considering that individuals use
different types of reference management tools such as
[Zotero](https://www.zotero.org/) and [Mendeley](https://www.mendeley.com/). To
simplify reference management on our blog and still allow blog members to use
their own preferred reference management tools, we decided to build a
centralized repo with a master `.bib` file and tools to update and validate this
file.


# Installation

The easiest way to install `inSilecoRef` is to use the
[`remotes`](https://CRAN.R-project.org/package=remotes) package:

```R
install.packages("remotes")
remotes::install_github("inSileco/inSilecoRef")
```

# Main features

The package offers two main functions. The main function, `newDOI()` uses
digital object identifier (DOI) and takes advantage of functionalities offered
by the [`rcrossref`](https://github.com/ropensci/rcrossrefs) and
[`bibtex`](https://github.com/ropensci/bibtex) packages.
The second function, `newManual()` allows users to create a manual entry in
cases where DOIs are not available.

## Update bibliography with DOI

```R
library(inSilecoRef)
add_(DOI = c('10.1016/j.tree.2016.10.011','10.1890/130230','10.1515/9781400881376'))
```

<br/>

```
R>>  The following DOI were not imported as they are already in the master .bib file:
R>>   10.1515/9781400881376
R>>
R>>  The following references were added to the master .bib file:
R>>  @article{Dee_2017,
R>>     doi = {10.1016/j.tree.2016.10.011},
R>>     url = {https://doi.org/10.1016%2Fj.tree.2016.10.011},
R>>     year = 2017,
R>>     month = {feb},
R>>     publisher = {Elsevier {BV}},
R>>     volume = {32},
R>>     number = {2},
R>>     pages = {118--130},
R>>     author = {Laura E. Dee and Stefano Allesina and Aletta Bonn and Anna Eklöf and Steven D. Gaines and Jes Hines and Ute Jacob and Eve McDonald-Madden and Hugh Possingham and Matthias Schröter and Ross M. Thompson},
R>>     title = {Operationalizing Network Theory for Ecosystem Service Assessments},
R>>     journal = {Trends in Ecology {\&} Evolution}
R>>  } @article{Low_D_carie_2014,
R>>     doi = {10.1890/130230},
R>>     url = {https://doi.org/10.1890%2F130230},
R>>     year = 2014,
R>>     month = {sep},
R>>     publisher = {Wiley},
R>>     volume = {12},
R>>     number = {7},
R>>     pages = {412--418},
R>>     author = {Etienne Low-D{\'{e}}carie and Corey Chivers and Monica Granados},
R>>     title = {Rising complexity and falling explanatory power in ecology},
R>>     journal = {Frontiers in Ecology and the Environment}
R>>  }
```

### To do list

- [ ] `manual.bib` manual entries bibtex file
- [ ] `newManual()` function to create new manual entry for `manual.bib`
- [ ] eventually add a function that updates bib file to identify which posts are citing which reference


NEWS.md
# inSilecoRef 0.1.1

* `add_doi_reference()` handle case sensitivity properly (DOI names are case insensitive).


# inSilecoRef 0.1.0 

* Re-write `newDOI()` that is now named `add_doi_reference()`.
* Review Actions workflows.
* Add new functions to quickly generate commonly used URL for R packages: `url_cran()`, `url_bioc()` and `url_gh()`.


# inSilecoRef 0.0.1

* Added a `NEWS.md` file to track changes to the package.
* Added newDOI.R


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(inSilecoRef)

test_check("inSilecoRef")


tests/testthat/test-url.R
test_that("CRAN URLs", {
    expect_identical(url_cran(""), "[CRAN](https://CRAN.R-project.org)")
    expect_identical(
        url_cran("fs"),
        "[`fs`](https://CRAN.R-project.org/package=fs)"
    )
    expect_error(url_cran("f!s"))
})

test_that("Bioconductor URLs", {
    expect_identical(
        url_bioc(""),
        "[Bioconductor](https://www.bioconductor.org/)"
    )
    expect_identical(
        url_bioc("ggtree"),
        "[`ggtree`](https://bioconductor.org/packages/ggtree/)"
    )
    expect_error(url_bioc("gg!"))
})

test_that("Bioconductor URLs", {
    expect_identical(
        url_bioc(""),
        "[Bioconductor](https://www.bioconductor.org/)"
    )
    expect_identical(
        url_bioc("ggtree"),
        "[`ggtree`](https://bioconductor.org/packages/ggtree/)"
    )
    expect_error(url_bioc("gg!"))
})

test_that("GitHub URLs", {
    expect_identical(
        url_gh(""),
        "[cran](https://github.com/cran)"
    )
    expect_identical(
        url_gh(user = ""),
        "[GitHub](https://github.com)"
    )
    expect_identical(
        url_gh("rcites"),
        "[`rcites`](https://github.com/cran/rcites)"
    )
    expect_identical(
        url_gh(user = "inSileco"),
        "[inSileco](https://github.com/inSileco)"
    )
    expect_identical(
        url_gh("rowlbot", "inSileco"),
        "[`rowlbot`](https://github.com/inSileco/rowlbot)"
    )
})


tests/testthat/test-doi-ref.R
bib <- "test.bib"
dois <- c(
    "10.1016/j.tree.2016.10.011",
    "10.1890/130230",
    "10.1515/9781400881376",
    "10.1007/BF01386390" # no case sensitivity
)

suppressMessages({
    test_that("add first doi", {
        res <- add_doi_reference(dois[1], bib)
        expect_equal(length(res), 1)
        expect_true(file.exists(bib))
        expect_true(inherits(res, "bibentry"))
    })

    test_that("add new doi", {
        res2 <- add_doi_reference(dois, bib)
        expect_equal(length(res2), 4)
        res3 <- add_doi_reference(dois[1], bib)
        expect_identical(res2, res3)
    })

    test_that("bad file format", {
        expect_error(add_doi_reference("10.1515/9781400881376", "test.b"))
    })
})
unlink(bib)

R/url_cran.R
#' URL for R packages
#'
#' Get an URL for a specific package (Markdown style).
#'
#' @param pkg a package name.
#' @param user a GitHub user name or a GitHub organisation name.
#'
#' @return
#' A link to the CRAN's canonical URL for a given package formatted in Markdown.
#' Note that is `pkg = ""` a link to the CRAN is returned.
#'
#' @export
#' @examples
#' url_cran("")
#' url_cran("ps")
#' url_gh("inSilecoRef", "inSileco")
#' url_bioc("CRImage")
url_cran <- function(pkg = "") {
  if (pkg == "") {
    "[CRAN](https://CRAN.R-project.org)"
  } else {
    stopifnot(is_package_name_valid(pkg))
    glue("[`{pkg}`](https://CRAN.R-project.org/package={pkg})")
  }
}

#' @describeIn url_cran A link to the Bioconductor's short URL for a given
#' package formatted in Markdown. Note that is `pkg = ""` a link to Bioconductor
#' is returned.
#' @export

url_bioc <- function(pkg = "") {
  if (pkg == "") {
    "[Bioconductor](https://www.bioconductor.org/)"
  } else {
    stopifnot(is_package_name_valid(pkg))
    glue("[`{pkg}`](https://bioconductor.org/packages/{pkg}/)")
  }
}

#' @describeIn url_cran A link to the GitHub repository for a given package.
#' If `pkg = ""` a link to the user account in returned and if `pkg = ""` and
# `user = ""` then a link to GitHub is returned.
#' @export

url_gh <- function(pkg = "", user = "cran") {
  if (pkg == "") {
    if (user == "") {
      "[GitHub](https://github.com)"
    } else {
      glue("[{user}](https://github.com/{user})")
    }
  } else {
    glue("[`{pkg}`](https://github.com/{user}/{pkg})")
  }
}


is_package_name_valid <- function(x) {
  grepl("^[[:alnum:]][[:alnum:]\\.]*[[:alnum:]]$", x)
}


R/add_doi_reference.R
#' Add DOI-based references to a bib file
#'
#' Check if DOIs are present in a bib file passed as an argument, if one or more
#' DOI are missing the associated references are imported and added to the
#' bib file.
#'
#' @param doi character vector of DOI to import.
#' @param bib character path to master .bib file, default is `inSilecoRef.bib`,
#' if the file does not exist it will be created.
#' @keywords DOI
#'
#' @export
#'
#' @examples
#' # Example 1:
#' \dontrun{
#' add_doi_reference(doi = c(
#'   "10.1016/j.tree.2016.10.011", "10.1890/130230",
#'   "10.1515/9781400881376"
#' ))
#' }
add_doi_reference <- function(doi, bib = "./inSilecoRef.bib") {
  # 1. Import .bib file
  if (!file.exists(bib)) {
    stopifnot(path_ext(bib) == "bib")
    cli_alert_info("Creating {bib}")
    file.create(bib)
    ref <- utils::bibentry()
  } else {
    cli_alert_info("Reading bib file {bib}")
    ref <- read.bib(bib)
  }

  # 2. Check wheteher dois are already in the bib file
  if (length(ref)) {
    # DOI names are case insensitive
    doi <- doi[!tolower(doi) %in% tolower(unlist(ref$doi))]
  }

  # 4. if !DOI %in% inSilecoRef.bib
  if (length(doi)) {
    cli_alert_info("Importing {length(doi)} DOI{?s}")
    new_ref <- rcrossref::cr_cn(dois = doi, format = "bibtex")
    tf <- tempfile()
    on.exit(unlink(tf))
    new_ref <- rcrossref::cr_cn(dois = doi, format = "bibtex") |>
      unlist() |>
      write(tf)
    new_ref <- read.bib(tf)
    id <- names(new_ref) %in% names(ref)
    if (length(id)) {
      names(new_ref)[id] <- rename_key(names(new_ref)[id])
    }
    cli_alert_info("Writing bib file {bib}")
    updated_ref <- c(ref, new_ref)
    write.bib(c(ref, new_ref), bib)
  } else {
    cli_alert_info("No reference to be added")
    updated_ref <- ref
  }

  invisible(return(updated_ref))
}


is_reference_empty <- function(x) {
  return(x == "" | is.null(x))
}

get_keys <- function(x) {
  out <- lapply(x, \(x) x$key)
  return(unlist(out))
}

get_dois <- function(x) {
  out <- lapply(x, \(x) x$doi)
  return(unlist(out))
}

rename_key <- function(x) {
  ukey <- unique(x)
  for (i in ukey) {
    x[x == i] <- paste0(x[x == i], ":", seq(x[x == i]))
  }
  return(x)
}

R/newDOI.R
#' Import and validate new DOI to DOI lists
#'
#' Imports and validates new entries for a DOI list
#'
#' @param DOI character vector of DOI to import
#' @param bib character path to master .bib file, default is `inSilecoRef.bib`
#'
#' @keywords DOI
#'
#' @export
#'
#' @examples
#' # Example 1:
#' \dontrun{
#' newDOI(DOI = c("10.1016/j.tree.2016.10.011", "10.1890/130230", "10.1515/9781400881376"))
#' }
newDOI <- function(DOI, bib = "./inSilecoRef.bib") {
  # Steps
  # 1. import inSilecoRef.bib
  # 2. Export list of inSilecoDOI from inSilecoRef.bib & year & family name of 1st author & key
  # 3. check DOI against inSilecoDOI
  # 4. if !DOI %in% inSilecoRef.bib
  #   4.1 import new references
  #   4.2 check and change keywords if duplicated
  # 5. Prompt user for confirmation of reference additions
  # 6. Reload and update inSilecoRef.bib as raw file

  # Function
  # 1. Import .bib file
  inSilecoRef <- RefManageR::ReadBib(bib)

  # 2. Export relevant info on bibliography
  nRef <- length(inSilecoRef)
  bibTab <- data.frame(
    DOI = character(nRef),
    key = character(nRef),
    family = character(nRef),
    year = character(nRef),
    stringsAsFactors = FALSE
  )

  for (i in seq_len(nRef)) {
    bibTab$DOI[i] <- inSilecoRef[i]$doi # DOI
    bibTab$key[i] <- names(inSilecoRef[i]) # key
    bibTab$family[i] <- inSilecoRef[i]$author[1L]$family # family name of 1st author
    bibTab$year[i] <- inSilecoRef[i]$year # year of publication
  }

  # 3. Check DOI against bibTab$DOI
  checkDOI <- DOI %in% bibTab$DOI

  # 4. if !DOI %in% inSilecoRef.bib
  if (sum(!checkDOI)) {
    # 4.1 Import reference using rcrossref
    newRef <- rcrossref::cr_cn(dois = DOI[!checkDOI], format = "bibtex")
    newRef <- unlist(newRef)

    # 4.2 Check key
    nNewRef <- length(newRef)
    for (i in seq_len(nNewRef)) {
      begChar <- str_locate(newRef[i], "(.*?)\\{")[2] + 1
      endChar <- str_locate(newRef[i], "(.*?),")[2] - 1
      key <- substr(newRef[i], begChar, endChar)

      # If key already exists, give a new key by adding a letter at the end
      if (key %in% bibTab$key) {
        nKey <- sum(str_detect(bibTab$key, key)) # Number of references with same key + letter
        newKey <- paste0(key, letters[nKey + 1])
        newRef[i] <- gsub(key, newKey, newRef[i])
      }
    }
  }

  # # 5. Report of changes made
  #   if (sum(checkDOI) > 0) {
  #     cat('The following DOI were not imported as they are already in the master .bib file: \n',
  #         paste0(DOI[checkDOI], '\n'))
  #     cat('\n')
  #   }
  #
  #   if (sum(!checkDOI) > 0) {
  #     cat('The following references were added to the master .bib file: \n')
  #     cat(newRef)
  #   }

  # 6. Reload and update inSilecoRef.bib as raw file
  # Thanks to https://stackoverflow.com/questions/9068397/import-text-file-as-single-character-string
  if (sum(!checkDOI) > 0) {
    bibRaw <- readChar(bib, file.info(bib)$size)
    updateRef <- c(bibRaw, newRef)
    write(updateRef, file = bib)
  }
}
